### PR TITLE
Adds an in game matrix tester so I can finish my work

### PR DESCRIPTION
**Pull request by tralezab**
### Very similar to filterrific, but just for matrixes. And a lot less time put into it. Because it's only for matrix math.

image

### tgui menu + demonstration:

image

## Why It's Good For The Game

Helps developers make interesting stuff. I mean, you don't have to merge it, I've gotten what i've needed for my own github project >:)

just thought it was useful so I pr'd it

## Changelog
:cl:
code: Added a matrix tester for setting matrix vectors and seeing how they affect the final product.
admin: You can probably find a way to abuse this!
/:cl:

### CODE REVIEW

**Comment by tralezab** [comment]
Now I will be able to finish #59081

**Comment by tralezab** [comment]
https://user-images.githubusercontent.com/40974010/118222712-f49a8b00-b434-11eb-93c4-99647105ed8d.mp4

File: code/modules/admin/view_variables/nobody_wants_to_learn_matrix_math.dm [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+/datum/nobody_wants_to_learn_matrix_math
+	var/atom/target
+	var/matrix/testing_matrix
+
+/datum/nobody_wants_to_learn_matrix_math/New(atom/target)
+	src.target = target
+	testing_matrix = new()
+
+/datum/nobody_wants_to_learn_matrix_math/Destroy(force, ...)
+	. = ..()
```
Comment (Mothblocks) [inline code/modules/admin/view_variables/nobody_wants_to_learn_matrix_math.dm:19]: Call `return ..()` last.

File: code/modules/client/client_procs.dm [hunk 2]
```diff
@@ -1098,6 +1098,10 @@ GLOBAL_LIST_INIT(blacklisted_builds, list(
 		holder.filteriffic = new /datum/filter_editor(in_atom)
 		holder.filteriffic.ui_interact(mob)
 
+/client/proc/open_matrix_tester(atom/in_atom)
+	if(holder)
+		var/datum/nobody_wants_to_learn_matrix_math/matrix_tester = new /datum/nobody_wants_to_learn_matrix_math(in_atom)
```
Comment (Mothblocks) [inline code/modules/client/client_procs.dm:1103]:
```suggestion
		var/datum/nobody_wants_to_learn_matrix_math/matrix_tester = new(in_atom)
```

File: code/game/atoms.dm [hunk 3]
```diff
@@ -1295,6 +1296,10 @@
 		var/client/C = usr.client
 		C?.open_filter_editor(src)
 
+	if(href_list[VV_HK_TEST_MATRIXES] && check_rights(R_VAREDIT))
+		var/client/C = usr.client
+		C?.open_matrix_tester(src)
```
Comment (Mothblocks) [inline code/game/atoms.dm:1301]:
```suggestion
		usr.client?.open_matrix_tester(src)
```

You don't need the cast, the other ones are dumb.

File: tgui/packages/tgui/interfaces/MatrixMathTester.js [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+                    step={0.005}
+                    format={value => toFixed(value, 2)}
+                    width="70px"
+                    onChange={(e, value) => act('change_var', {
+                      var_name: "c",
+                      var_value: value,
+                    })} />
+                </Stack.Item>
+                <Stack.Item grow>
+                  <NumberInput
```
Comment (Mothblocks) [inline tgui/packages/tgui/interfaces/MatrixMathTester.js:102]: You repeat this a lot, I think you could probably make it a small component.

**Review by Mothblocks — CHANGES_REQUESTED** [review]

File: tgui/packages/tgui/interfaces/MatrixMathTester.js (hunk 4 above)
Comment (tralezab) [reply tgui/packages/tgui/interfaces/MatrixMathTester.js:102]: *soft sobbing noises*

Comment (tralezab) [reply tgui/packages/tgui/interfaces/MatrixMathTester.js:102]: don't do this to me...................

Comment (tralezab) [reply tgui/packages/tgui/interfaces/MatrixMathTester.js:102]: oh wait that's not that hard

Comment (Mothblocks) [reply tgui/packages/tgui/interfaces/MatrixMathTester.js:102]: dumby

File: code/__DEFINES/vv.dm [hunk 5]
```diff
@@ -87,6 +87,7 @@
 #define VV_HK_AUTO_RENAME "auto_rename"
 #define VV_HK_RADIATE "radiate"
 #define VV_HK_EDIT_FILTERS "edit_filters"
+#define VV_HK_TEST_MATRIXES "test_matrixes"
```
Comment (stylemistake) [inline code/__DEFINES/vv.dm:90]: It's matrices?

Comment (tralezab) [reply code/__DEFINES/vv.dm:90]: True!

**Comment by stylemistake** [comment]
Why is this UI so huge

**Comment by tralezab** [comment]
...I'm gonna need to give the tgui a second pass, lol

**Comment by tralezab** [comment]
Someone can pick this up if they think it's worth adding, I kinda sobered up about the whole thing and i'm personally whatever about it now so

**Pull request closed without merging**